### PR TITLE
Cleanup QueryTest framework and be more strict

### PR DESCRIPTION
**Pull request by s1monw**
This commit contains:
- renaming BaseQueryTestCase to AbstractQueryTestCase
- uses always STRICT parsing when parsing from builders XContent
- ensures that SearchContext is only but always available during toQuery but never during parse phase
- adds a way to override the default ParseFieldMatcher to allow queries to set deprecated API by default

this also found a bug

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/NotQueryParser.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
  */
 public class NotQueryParser extends BaseQueryParser<NotQueryBuilder> {
 
-    private static final ParseField QUERY_FIELD = new ParseField("filter", "query");
+    private static final ParseField QUERY_FIELD = new ParseField("query", "filter");
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/NotQueryParser.java]: nasty... we have to backport this to 2.x and 2.0, and master if not filter is still there.

File: core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java [hunk 2]
```diff
@@ -254,7 +250,7 @@ public void testFromXContent() throws IOException {
         QB testQuery = createTestQueryBuilder();
         assertParsedQuery(testQuery.toString(), testQuery);
         for (Map.Entry<String, QB> alternateVersion : getAlternateVersions().entrySet()) {
-            assertParsedQuery(alternateVersion.getKey(), alternateVersion.getValue());
+            assertParsedQuery(alternateVersion.getKey(), alternateVersion.getValue(), ParseFieldMatcher.EMPTY);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java]: I think this should be strict too. The problem is that the not query has some alternate version that uses deprecated syntax. I think that should be moved to a separate test like you did with recent PRs.

**Comment by javanna** [comment]
this looks great, thanks a lot @s1monw . left one comment.

File: core/src/main/java/org/elasticsearch/index/query/NotQueryParser.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/query/NotQueryParser.java]: yeah man that's what I wanted to catch with this!!

**Comment by s1monw** [comment]
@javanna I cleaned it up even further and use strict parsing everywhere. can you take another look?

**Comment by javanna** [comment]
looks perfect, thanks a lot!

**Pull request merged**